BlazarService/src/main/java/com/hubspot/blazar/listener/SlackRoomNotificationVisitor.java

```diff
@@ -60,13 +62,57 @@ public void visit(RepositoryBuild build) throws Exception {
     Set<InstantMessageConfiguration> configurationSet = instantMessageConfigurationService.getAllWithBranchId(build.getBranchId());
     Optional<RepositoryBuild> previous = repositoryBuildService.getPreviousBuild(build);
     for (InstantMessageConfiguration instantMessageConfiguration : configurationSet) {
-      if (shouldSend(instantMessageConfiguration, build.getState(), previous, build)) {
-        sendSlackMessage(instantMessageConfiguration, build);
+      if (shouldSendRepositoryBuild(instantMessageConfiguration, build.getState(), previous, build)) {
+        sendSlackMessageWithRetries(instantMessageConfiguration.getChannelName(), slackUtils.buildSlackAttachment(build));
       }
     }
   }
 
-  private boolean shouldSend(InstantMessageConfiguration instantMessageConfiguration, RepositoryBuild.State state, Optional<RepositoryBuild> previous, RepositoryBuild build) {
+  @Override
+  public void visit(ModuleBuild build) throws Exception {
+    if (!(build.getState().isComplete())) {
+      return;
+    }
+    Set<InstantMessageConfiguration> configurationSet = instantMessageConfigurationService.getAllWithModuleId(build.getModuleId());
+    Optional<ModuleBuild> previous = moduleBuildService.getPreviousBuild(build);
+    for (InstantMessageConfiguration instantMessageConfiguration : configurationSet) {
+      if (shouldSendModuleBuild(instantMessageConfiguration, build.getState(), previous, build)) {
+        sendSlackMessageWithRetries(instantMessageConfiguration.getChannelName(), slackUtils.buildSlackAttachment(build));
+      }
+    }
+  }
+
+  private void sendSlackMessageWithRetries(String channelName, SlackAttachment attachment) {
+    try {
+      SlackUtils.makeSlackMessageSendingRetryer().call(() -> sendSlackMessage(channelName, attachment));
+      metricRegistry.counter("successful-slack-channel-sends").inc();
+    } catch (Exception e) {
+      metricRegistry.counter("failed-slack-channel-sends").inc();
+      LOG.error("Could not send slack message {}", attachment, e);
+    }
+  }
+
+  private boolean sendSlackMessage(String channelName, SlackAttachment attachment) {
+    if (!slackSession.isPresent()) {
+      LOG.debug("No slack session present, not sending message {} to channel {}", attachment.toString(), channelName);
+      return true;
+    }
+
+    Optional<SlackChannel> slackChannel = Optional.fromNullable(slackSession.get().findChannelByName(channelName));
+    if (!slackChannel.isPresent()) {
+      LOG.warn("No slack channel found for name {}", channelName);
+      return false;
+    }
+
+    SlackMessageHandle<SlackMessageReply> result = slackSession.get().sendMessage(slackChannel.get(), "", attachment);
+    if (!result.getReply().isOk()) {
```

Comment:
if (result ==null || !result.getReply().isOk())

if you check the code of the library it is not handling exceptions that gracefully:
```catch (Exception e)
        {
            // TODO : improve exception handling
            e.printStackTrace();
        }```
so you can end up with a null reply